streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -568,7 +577,9 @@ private void runLoop() {
         }
 
         if (rebalanceException != null) {
-            throw new StreamsException(logPrefix + " Failed to rebalance.", rebalanceException);
+            if (!(rebalanceException instanceof ProducerFencedException)) {
```

Comment:
I do not fully understand the `rebalanceException` usage here: we always rethrow the exception after setting it in the callback, then why do we want to wrap it again as `StreamsException` here?

More generally speaking, I'd like to use this opportunity to make a pass over all the places that we can throw an exception and re-visit the manner how we capture / rethrow them. Personally I felt that:

1. Any exceptions thrown from embedded consumer, like `conusmer.poll` should be captured and handled without rethrowing to the user, and depending on the type of the thrown exception we may a) shutdown the whole instance b) close corresponding tasks.

2. Any exceptions thrown from embedded producer, like `producer#xxTxn, #send, #flush` etc should be capture and handled without rethrowing to the user. Similar handling logic should apply as above.

3. Any exceptions throwable from the streams layer should be captured, logged, and then some of them should be re-throw to the user (e.g. putRecords which may call serde, io exception from writing offset file, rocksdb store etc) while others should be handled without re-throwing (e.g. offset commit failed, timeout on fetching metadata inside changlog reader, etc).

4. Any exceptions thrown from user code (processor#process, punctuate, close, store#put, get, flush) should be captured and rethrow to the users. And in the future we would further categorize them to finer exception handling than all-the-way to global uncaught exception handler (cc @enothereska )

I understand this would make this PR a bit longer to review, but without explicitly making a pass over all the exceptions we are handling it is difficult to see if all error cases are well handled when EOS is turned on.